File: go/httpmw/options.go
package httpmw

import (
	"net/http"
	"strings"

	"github.com/sns45/anyonce/go/anyonce"
	"github.com/sns45/anyonce/go/internal/httpx"
)

// FingerprintMode selects the D9 fingerprint: body hashes method, path with query and the raw bytes; jcs hashes the
// RFC 8785 form of a JSON body instead of the bytes and falls back to body when the payload is not JSON.
type FingerprintMode string

// FingerprintBody and FingerprintJCS are the two D9 fingerprint modes.
const (
	FingerprintBody FingerprintMode = "body"
	FingerprintJCS  FingerprintMode = "jcs"
)

// DefaultHeaderName, DefaultMaxRequestBytes and DefaultProblemBaseURI are the Options defaults (REQ-HTTP-2,
// REQ-HTTP-6, REQ-HTTP-13). The last two are shared with the other doors and live in internal/httpx.
const (
	DefaultHeaderName      = "Idempotency-Key"
	DefaultMaxRequestBytes = httpx.DefaultMaxRequestBytes
	DefaultProblemBaseURI  = httpx.DefaultProblemBaseURI
)

// DefaultMethods are the methods the layer applies to (REQ-HTTP-1).
var DefaultMethods = []string{http.MethodPost, http.MethodPatch}

// DefaultStoreHeaders is the REQ-HTTP-8 allowlist. Set-Cookie is never stored whatever the allowlist says.
var DefaultStoreHeaders = []string{"Content-Type", "Content-Language", "Location", "ETag", "Link"}

// Options configures the middleware. Zero values take the documented defaults.
type Options struct {
	// Methods the layer applies to; others pass through (REQ-HTTP-1). Default POST and PATCH.
	Methods []string
	// HeaderName is matched case-insensitively (REQ-HTTP-2). Default Idempotency-Key.
	HeaderName string
	// Required makes a missing header a 400 missing-key; otherwise the request passes through (REQ-HTTP-3).
	Required bool
	// KeySyntax is D7: lenient (default) or strict.
	KeySyntax anyonce.Syntax
	// Scope replaces the default scope of METHOD plus path, where the path is the escaped form (REQ-HTTP-5). The
	// middleware wraps the mux, so no route pattern is known here; pass one from your router if you want
	// pattern-level scopes.
	Scope func(*http.Request) string
	// Principal is appended to the scope after a hash; an empty string means none (REQ-HTTP-5).
	Principal func(*http.Request) string
	// RequirePrincipal makes an empty principal a 500 missing-principal (Q18). New panics when it is set without Principal.
	RequirePrincipal bool
	// Fingerprint mode (D9). Default body. FingerprintFunc overrides both modes.
	Fingerprint     FingerprintMode
	FingerprintFunc func(*http.Request, []byte) (string, error)
	// MaxRequestBytes bounds the body the layer reads; larger bodies are 413 (REQ-HTTP-6). Default 1 MiB.
	MaxRequestBytes int64
	// StoreHeaders is the response header allowlist (REQ-HTTP-8).
	StoreHeaders []string
	// Policy is the engine policy (lease, TTL, cap, StoreResult, OnStoreError, Clock, Hooks). Zero values take
	// anyonce.DefaultPolicy values.
	Policy anyonce.Policy
	// ProblemBaseURI is D11. DocsURL is the Link target on a 400 missing-key; default ProblemBaseURI plus missing-key.
	ProblemBaseURI string
	DocsURL        string
	// OnError renders a problem differently (REQ-HTTP-13). Status and code must not change.
	OnError func(w http.ResponseWriter, r *http.Request, p Problem)
	// Skip opts a request out (REQ-HTTP-15).
	Skip func(*http.Request) bool
}

// resolved is Options with every default applied and lookup sets precomputed.
type resolved struct {
	Options
	methods      map[string]bool
	storeHeaders map[string]bool
}

// resolve applies the documented defaults and builds the methods and storeHeaders lookup sets.
func (o Options) resolve() resolved {
	r := resolved{Options: o, methods: map[string]bool{}, storeHeaders: map[string]bool{}}
	methods := o.Methods
	if len(methods) == 0 {
		methods = DefaultMethods
	}
	for _, m := range methods {
		r.methods[strings.ToUpper(m)] = true
	}
	headers := o.StoreHeaders
	if headers == nil {
		headers = DefaultStoreHeaders
	}
	for _, h := range headers {
		r.storeHeaders[http.CanonicalHeaderKey(h)] = true
	}
	if r.HeaderName == "" {
		r.HeaderName = DefaultHeaderName
	}
	if r.KeySyntax == "" {
		r.KeySyntax = anyonce.SyntaxLenient
	}
	if r.Fingerprint == "" {
		r.Fingerprint = FingerprintBody
	}
	if r.MaxRequestBytes <= 0 {
		r.MaxRequestBytes = DefaultMaxRequestBytes
	}
	if r.ProblemBaseURI == "" {
		r.ProblemBaseURI = DefaultProblemBaseURI
	}
	if r.DocsURL == "" {
		r.DocsURL = r.ProblemBaseURI + string(CodeMissingKey)
	}
	return r
}

File: packages/core/src/http/options.ts
import {
  DEFAULT_MAX_RESULT_BYTES,
  defaultPolicy,
  type ExecuteHooks,
  type ExecutePolicy,
} from '../engine';
import type { KeySyntax } from '../key';
import type { Store, StoredResult } from '../types';
import { DEFAULT_PROBLEM_BASE_URI, type Problem } from './problems';

export type FingerprintMode = 'body' | 'jcs';
export type FingerprintFn = (req: Request, body: Uint8Array) => Promise<string> | string;

export const DEFAULT_METHODS = ['POST', 'PATCH'];
export const DEFAULT_HEADER_NAME = 'Idempotency-Key';
export const DEFAULT_MAX_REQUEST_BYTES = 1_048_576;
export const DEFAULT_STORED_HEADERS = [
  'Content-Type',
  'Content-Language',
  'Location',
  'ETag',
  'Link',
];

/** Options for withIdempotency and the framework bindings (requirements 4.4). */
export interface HttpIdempotencyOptions {
  store: Store;
  /** REQ-HTTP-1: methods the layer applies to; others pass through. Default POST and PATCH. */
  methods?: string[];
  /** REQ-HTTP-2: header name, matched case-insensitively. Default Idempotency-Key. */
  headerName?: string;
  /** REQ-HTTP-3: a missing header is 400 missing-key when true, pass-through when false. Default false. */
  required?: boolean;
  /** REQ-HTTP-4 and D7. Default lenient. */
  keySyntax?: KeySyntax;
  /** REQ-HTTP-5: replaces the default scope of METHOD plus route pattern or pathname. */
  scope?: (req: Request) => string;
  /** REQ-HTTP-5: appended to the scope after a hash. */
  principal?: (req: Request) => string | undefined;
  /** REQ-HTTP-5: a principal function that returns nothing yields 500 missing-principal. Default false. */
  requirePrincipal?: boolean;
  /** REQ-HTTP-6 and D9. Default body. */
  fingerprint?: FingerprintMode | FingerprintFn;
  /** REQ-HTTP-6: larger bodies are 413 payload-too-large. Default 1 MiB. */
  maxRequestBytes?: number;
  /** REQ-HTTP-8: response headers stored and replayed. Set-Cookie is never stored. */
  storeHeaders?: string[];
  leaseMs?: number;
  ttlMs?: number;
  /**
   * D12: larger results are stored in the omitted form. Default 1 MiB, and never more than the store's own
   * `maxResultBytes` when it declares one (Q20), so a store that cannot hold a result is never asked to.
   */
  maxResultBytes?: number;
  storeResult?: (result: StoredResult) => boolean;
  onStoreError?: 'fail-closed' | 'fail-open';
  clock?: () => number;
  hooks?: ExecuteHooks;
  hookErrors?: { count: number };
  /** D11. Default https://in8.sh/anyonce/problems/ */
  problemBaseUri?: string;
  /** REQ-HTTP-3: the Link target on a 400 missing-key. Default problemBaseUri plus missing-key. */
  docsUrl?: string;
  /** REQ-HTTP-13: render a problem differently. Status and code must not change. */
  onError?: (problem: Problem, req: Request) => Response | Promise<Response>;
  /** REQ-HTTP-15: per-request opt-out. */
  skip?: (req: Request) => boolean;
}

export interface ResolvedHttpOptions {
  store: Store;
  methods: Set<string>;
  headerName: string;
  required: boolean;
  keySyntax: KeySyntax;
  scope?: (req: Request) => string;
  principal?: (req: Request) => string | undefined;
  requirePrincipal: boolean;
  fingerprint: FingerprintMode | FingerprintFn;
  maxRequestBytes: number;
  storeHeaders: Set<string>;
  policy: ExecutePolicy;
  problemBaseUri: string;
  docsUrl: string;
  onError?: (problem: Problem, req: Request) => Response | Promise<Response>;
  skip?: (req: Request) => boolean;
}

export function resolveHttpOptions(options: HttpIdempotencyOptions): ResolvedHttpOptions {
  if (options.requirePrincipal === true && options.principal === undefined) {
    throw new TypeError('anyonce: requirePrincipal is true but no principal function was given');
  }
  const policyOverrides: Partial<ExecutePolicy> = {};
  if (options.leaseMs !== undefined) policyOverrides.leaseMs = options.leaseMs;
  if (options.ttlMs !== undefined) policyOverrides.ttlMs = options.ttlMs;
  // Q20: the policy cap never exceeds what the store says it can hold whole.
  policyOverrides.maxResultBytes = Math.min(
    options.maxResultBytes ?? DEFAULT_MAX_RESULT_BYTES,
    options.store.maxResultBytes ?? Number.POSITIVE_INFINITY,
  );
  if (options.storeResult !== undefined) policyOverrides.storeResult = options.storeResult;
  if (options.onStoreError !== undefined) policyOverrides.onStoreError = options.onStoreError;
  if (options.clock !== undefined) policyOverrides.clock = options.clock;
  if (options.hooks !== undefined) policyOverrides.hooks = options.hooks;
  if (options.hookErrors !== undefined) policyOverrides.hookErrors = options.hookErrors;
  const problemBaseUri = options.problemBaseUri ?? DEFAULT_PROBLEM_BASE_URI;
  const resolved: ResolvedHttpOptions = {
    store: options.store,
    methods: new Set((options.methods ?? DEFAULT_METHODS).map((m) => m.toUpperCase())),
    headerName: options.headerName ?? DEFAULT_HEADER_NAME,
    required: options.required ?? false,
    keySyntax: options.keySyntax ?? 'lenient',
    requirePrincipal: options.requirePrincipal ?? false,
    fingerprint: options.fingerprint ?? 'body',
    maxRequestBytes: options.maxRequestBytes ?? DEFAULT_MAX_REQUEST_BYTES,
    storeHeaders: new Set(
      (options.storeHeaders ?? DEFAULT_STORED_HEADERS).map((h) => h.toLowerCase()),
    ),
    policy: defaultPolicy(policyOverrides),
    problemBaseUri,
    docsUrl: options.docsUrl ?? `${problemBaseUri}missing-key`,
  };
  if (options.scope !== undefined) resolved.scope = options.scope;
  if (options.principal !== undefined) resolved.principal = options.principal;
  if (options.onError !== undefined) resolved.onError = options.onError;
  if (options.skip !== undefined) resolved.skip = options.skip;
  return resolved;
}
